### PR TITLE
ISSUE-115: fix bug where 'value' would be cached for dynamic fields.

### PR DESCRIPTION
**Pull request by laktech**

### CODE REVIEW

**Review by lon-pilot-mpf — APPROVED** [review]

File: iabtcf-decoder/src/main/java/com/iabtcf/utils/FieldDefs.java [hunk 1]
```diff
@@ -230,38 +229,42 @@ public int getEnd(BitReader bbv) {
...
-
-                return value;
-            } else {
-                return value;
             }
+
+            return value;
         }
 
         private boolean isDynamicPvt() {
```
Comment (srinivas81) [inline iabtcf-decoder/src/main/java/com/iabtcf/utils/FieldDefs.java:263]: Shouldn't this method be synchronized? Or at least use something like an AtomicBoolean with a getAndSet atomic operation?

File: iabtcf-extras-jackson/src/test/java/com/iabtcf/extras/jackson/gvl/VendorTest.java [hunk 2]
```diff
@@ -112,6 +112,6 @@ public void testGetOverflow() {
 
     @Test
     public void testIsDeleted() {
-        Assert.assertFalse(vendorEight.isDeleted());
+        Assert.assertTrue(vendorEight.isDeleted());
```
Comment (srinivas81) [inline iabtcf-extras-jackson/src/test/java/com/iabtcf/extras/jackson/gvl/VendorTest.java:115]: Is this change related to something else?

File: iabtcf-decoder/src/main/java/com/iabtcf/decoder/TCStringV2.java [hunk 3]
```diff
@@ -9,9 +9,9 @@
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
  * You may obtain a copy of the License at
- * 
+ *
  *      http://www.apache.org/licenses/LICENSE-2.0
- * 
+ *
```
Comment (srinivas81) [inline iabtcf-decoder/src/main/java/com/iabtcf/decoder/TCStringV2.java:14]: Unnecessary change?

File: iabtcf-decoder/src/main/java/com/iabtcf/utils/FieldDefs.java (hunk 1 above)
Comment (laktech) [reply iabtcf-decoder/src/main/java/com/iabtcf/utils/FieldDefs.java:263]: i don't think so, what are you concerned about?

File: iabtcf-extras-jackson/src/test/java/com/iabtcf/extras/jackson/gvl/VendorTest.java (hunk 2 above)
Comment (laktech) [reply iabtcf-extras-jackson/src/test/java/com/iabtcf/extras/jackson/gvl/VendorTest.java:115]: a failing test

File: iabtcf-decoder/src/main/java/com/iabtcf/decoder/TCStringV2.java (hunk 3 above)
Comment (laktech) [reply iabtcf-decoder/src/main/java/com/iabtcf/decoder/TCStringV2.java:14]: extraneous whitespace was removed

**Comment by SleimanJneidi** [comment]
why this needs to be thread-safe?

**Comment by laktech** [comment]
> why this needs to be thread-safe?

An issue was submitted that TCStringDecoder#decode fails under multi-threaded situations. Since it's a static method, it's not easy to create a per-thread instance. It seems like a reasonable thing to do.

File: iabtcf-decoder/src/main/java/com/iabtcf/utils/FieldDefs.java (hunk 1 above)
Comment (srinivas81) [reply iabtcf-decoder/src/main/java/com/iabtcf/utils/FieldDefs.java:263]: Sorry, wrong place; I meant at line 192. I will write the comment at the proper line.

File: iabtcf-decoder/src/main/java/com/iabtcf/utils/FieldDefs.java [hunk 4]
```diff
@@ -194,7 +194,6 @@ protected boolean isDynamic() {
             isDynamic = offset.isDynamic() || length.isDynamic();
             isDynamicInit = true;
         }
-
```
Comment (srinivas81) [inline iabtcf-decoder/src/main/java/com/iabtcf/utils/FieldDefs.java:197]: Should this method be synchronized as the comments (on the abstract method) say it should be thread safe. Do you think it's ok for two threads to enter the if loop, which I think is possible in this implementation.

File: iabtcf-extras-jackson/src/test/java/com/iabtcf/extras/jackson/gvl/VendorTest.java (hunk 2 above)
Comment (srinivas81) [reply iabtcf-extras-jackson/src/test/java/com/iabtcf/extras/jackson/gvl/VendorTest.java:115]: Interesting, I wonder how I didn't notice this earlier.

File: iabtcf-decoder/src/main/java/com/iabtcf/utils/FieldDefs.java (hunk 4 above)
Comment (laktech) [reply iabtcf-decoder/src/main/java/com/iabtcf/utils/FieldDefs.java:197]: This method is thread safe, it doesn't need to be synchronized.

Comment (laktech) [reply iabtcf-decoder/src/main/java/com/iabtcf/utils/FieldDefs.java:197]: to be more clear, it doesn't matter if multiple threads concurrently execute this method. the result will be the same. no need for additional synchronization constructs.

Comment (srinivas81) [reply iabtcf-decoder/src/main/java/com/iabtcf/utils/FieldDefs.java:197]: Sounds good.

**Review by srinivas81 — APPROVED** [review]

**Comment by SleimanJneidi** [comment]
> > why this needs to be thread-safe?
> 
> An issue was submitted that TCStringDecoder#decode fails under multi-threaded situations. Since it's a static method, it's not easy to create a per-thread instance. It seems like a reasonable thing to do.

TCStringDecoder#decode will always create a new instance of TCString. I doubt that the issue is threading, and I doubt this solves the hypothetical issue

**Comment by laktech** [comment]
> > > why this needs to be thread-safe?
> > 
> > 
> > An issue was submitted that TCStringDecoder#decode fails under multi-threaded situations. Since it's a static method, it's not easy to create a per-thread instance. It seems like a reasonable thing to do.
> 
> TCStringDecoder#decode will always create a new instance of TCString. I doubt that the issue is threading, and I doubt this solves the hypothetical issue

Th issue is not with TCString. TCString is not multi-threaded and the issue is not asking for it to be. The issue is with TCStringDecoder#decode failing in a multi-threaded way because of global state.
